src/dpl/src/OptMirror.cpp

```diff
@@ -143,28 +159,177 @@ Opendp::findMirrorCandidates(NetBoxes &net_boxes,
 }
 
 int
-Opendp::mirrorCandidates(vector<dbInst*> &mirror_candidates)
-{
+Opendp::mirrorCandidates(vector<dbInst*> &mirror_candidates,
+                          map<dbNet*, Rect> &net2box) {
   int mirror_count = 0;
   for (dbInst *inst : mirror_candidates) {
     // Use hpwl of all nets connected to the instance terms
     // before/after to determine incremental change to total hpwl.
-    int64_t hpwl_before = hpwl(inst);
     dbOrientType orient = inst->getOrient();
     dbOrientType orient_my = orientMirrorY(orient);
-    inst->setLocationOrient(orient_my);
-    int64_t hpwl_after = hpwl(inst);
-    if (hpwl_after > hpwl_before)
-      // Undo mirroring if hpwl is worse.
-      inst->setLocationOrient(orient);
-    else {
-      //printf("mirror %s\n", inst->getConstName());
+    int pt_x, pt_y;
+    inst->getLocation(pt_x, pt_y);
+    vector<pair<dbNet*, Rect*>> updated_box;
+    int64_t delta = hpwlIncrement(inst, pt_x, true, net2box, updated_box);
+
+    if (delta <= 0) {
       mirror_count++;
+      inst->setLocationOrient(orient_my);
+      for (auto item : updated_box) {
+        net2box[item.first].reset(item.second->xMin(), item.second->yMin(),
+                                  item.second->xMax(), item.second->yMax());
+      }
     }
   }
   return mirror_count;
 }
 
+int64_t
+Opendp::hpwl(dbInst *inst)
+{
+  int64_t inst_hpwl = 0;
+  for (dbITerm *iterm : inst->getITerms()) {
+    dbNet *net = iterm->getNet();
+    if (net)
+      inst_hpwl += hpwl(net);
+  }
+  return inst_hpwl;
+}
+
+int64_t
+Opendp::hpwlIncrement(dbInst *inst, int pt_x, bool mirror,
+                      map<dbNet*, Rect> &net2box,
+                      vector<pair<dbNet*, Rect*>> &updated_box) {
+  int64_t delta = 0;
+
+  // Net to insterm mapping
+  map<dbNet*, vector<dbITerm *>> net2iterm;
+
+  for (dbITerm *iterm : inst->getITerms()) {
+    if (iterm->getNet() != NULL) {
+      dbNet *net = iterm->getNet();
+
+      if (net2iterm.find(net) == net2iterm.end()) {
+        net2iterm[net] = vector<dbITerm *>();
+        net2iterm[net].push_back(iterm);
+      } else {
+        net2iterm[net].push_back(iterm);
+      }
+    }
+  }
+
+  for ( auto &item : net2iterm ) {
+    dbNet *net = item.first;
+    if (!isSupply(net)) {
+      Rect *new_net_box = new Rect();
+      delta += hpwlIncrementHelper(inst, item.second, net, pt_x, mirror,
+                                  new_net_box, net2box[net]);
+      updated_box.push_back({net, new_net_box});
+    }
+  }
+  return delta;
+}
+
+int64_t
+Opendp::hpwlIncrementHelper(dbInst *inst, vector<dbITerm *> &iterms,
```

Comment:
the helper could use a helper. this is too long for one function

Reply:
When we flip a cell
1. If the terminal's initial and final location is inside the net bbox, then the net's HPWL won't change.
2. If the final location is outside the net bbox, then the updated net bbox is old net bbox expanded till the final location.
3. If the initial location of the terminal is on the net bbox boundary and the final location is inside the net bbox, then we have to compute the final net bbox by traversing all the connected terminals of the net.

I have added another helper function for the third scenario in the latest commit.